eeprom.c

```diff
@@ -6,45 +6,65 @@
  * Licensed under the GPL-2.
  */
 
-#include <ftw.h>
-#include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
 #include <libgen.h>
 
 #include "eeprom.h"
 
-static char *eeprom_path = NULL;
-
-/* Test if a given file is a valid XCOMM EEPROM file. */
-static int is_eeprom(const char *fpath, const struct stat *sb,
-		int typeflag, struct FTW *ftwbuf)
-{
-	int ret = 0;
-	char *fpath_s = strdup(fpath);
-	if (typeflag == FTW_F && !strcmp(basename(fpath_s), "eeprom") \
-			&& sb->st_size == FAB_SIZE_FRU_EEPROM) {
-
-		eeprom_path = strdup(fpath);
-		ret = 1;
-	}
-	free(fpath_s);
-	return ret;
-}
 
 /* Recursively search a given path (defaulting to /sys) for an XCOMM compatible
  * EEPROM file.
  *
  * If a matching EEPROM file is found the path is returned, otherwise returns
  * NULL. Note that the string for the returned path is obtained with malloc and
- * should be freed.
+ * should be freed. It's up to the caller to free the allocated moemroy.
  */
 const char *find_eeprom(const char *path)
 {
-	if (path == NULL)
+	char *eeprom_path = NULL;
+	char eeprom_names[512];
+	FILE *fp = NULL;
+	char cmd[512];
+
+	if (path == NULL) {
 		path = "/sys";
+	}
+
+	snprintf(cmd, sizeof(cmd), "find %s -name eeprom 2>/dev/null", path);
+
+	fp = popen(cmd, "r");
+	if (fp == NULL) {
+		perror("popen");
+		return NULL;
+	}
+
+	while (fgets(eeprom_names, sizeof(eeprom_names), fp) != NULL) {
+		struct stat eeprom_file;
+		char *__basename = NULL;
+
+		if (eeprom_names[strlen(eeprom_names) - 1] == '\n')
+			eeprom_names[strlen(eeprom_names) - 1] = '\0';
+
+		__basename = strdup(eeprom_names);
```

Comment:
`strdup()` -ing `eeprom_names` seems un-necessary if you are not modifying it & free-ing it afterwards

Reply:
Im doing it because im calling `basename()` afterwards which will modify the passed string (in this case `eeprom_names`).

Reply:
ack